### PR TITLE
A Docker-based interactive Jupyterlab powered by GPU for artificial intelligence in Galaxy

### PR DESCRIPTION
**Pull request by anuprulez**
This PR adds a training material for image segmentation prediction using COVID CT scans in a Jupyterlab notebook. The Jupyterlab notebook is integrated into Galaxy Europe as an interactive tool. The standard Jupyterlab has been augmented with many features such as Git integration, Elyra AI, GPU support and runs inside a [Docker container](https://github.com/anuprulez/ml-jupyter-notebook). Moreover, remote processing of long-running deep learning training tasks is also possible via running a Galaxy tool that uploads scripts and datasets to Galaxy history and executes scripts on a remote cluster. The entire analysis comes from the paper: https://www.sciencedirect.com/science/article/pii/S2666990021000069 

Ground truth and predicted COVID CT scan masks:
true_pred_masks

Remote processing:
finished_history_remote_ai

3D protein structure (for a shorter SARS-CoV-2 spike protein sequence of length 300)

3D_300_L_protein

A bit of work is still left for this tutorial such as the addition of another use-case of integration with AlphaFold2. Would love to get some comments from @kxk302 and others as well. Thanks!!

Edit on 25.03.22: It is ready to be reviewed.

ping @bgruening

### CODE REVIEW

**Comment by anuprulez** [comment]
This PR is ready to be reviewed. @kxk302 would you have time to review this? Thanks!

**Comment by kxk302** [comment]
Hi @anuprulez 
Yes, will do.

**Comment by kxk302** [comment]
Sorry @anuprulez, did not get to review this today. Will review on Monday!
Have a great weekend.

**Comment by anuprulez** [comment]
@kxk302 no worries, have a nice weekend

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
+---
+layout: tutorial_hands_on
+
+title: A Docker-based interactive Jupyterlab powered by GPU for artificial intelligence in Galaxy
+zenodo_link: https://zenodo.org/record/6091361#.Ygu4gIzMI5k
```
Comment (hexylena) [inline topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:5]:
```suggestion
zenodo_link: https://zenodo.org/record/6091361
```

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+- machine-learning
+- deep-learning
+- jupyter-lab
+- image-segmentation
+contributors:
+- anuprulez
+
+---
+
+{:.no_toc}
```
Comment (hexylena) [inline topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:38]:
```suggestion
```

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
...dard model format, **open neural network exchange (ONNX)**, has been supported in the notebook to transform scikit-learn and tensorflow models to "onnx" files. These model files can be conveniently shared and used for inference. Galaxy also supports "onnx" file format which makes it easier to interoperate such models from a notebook to Galaxy and vice-versa. Using Bioblend, the notebook can be **connected to Galaxy** and its histories, tools and workflows can be accessed directly in the notebook. They can easily be operated via commands from the notebook. Many notebooks are created serving different purposes. These notebooks can be knit together to form one pipeline where each notebook transforms data taking a different form of data from its previous notebook and passing on the transformed data to its next notebook. An example workflow created using **[Elyra AI](https://elyra.readthedocs.io/en/stable/)** can be found at [<<open notebook>>/elyra/METABRIC_ML.pipeline]. These pipelines can also be executed remotely on different clusters too using its "runtimes" features that pull different docker containers. An example pipeline created in the notebook can be seen in the picture below.
```
Comment (hexylena) [inline topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:54]: `[<<open notebook>>/elyra/METABRIC_ML.pipeline]`

looks odd here

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+![Unet architecture](../../images/Unet.jpg "Architecture of Unet neural network for image segmentation. The name "Unet" resembles the shape of its architecture as 'U'")
+
+In this tutorial, we will use the dataset of CTs scans and their respective masks to train an Unet neural network model. The model learns to map the infected regions in the CT scans to their masks. For prediction, the trained model is given unseen CT scans and for each CT scan, it predicts infected regions or masks. For this experiment, we will use Jupyterlab for cloning the notebooks from Github that contain scripts for downloading data, creating and training Unet model and prediction tasks. Data (CT scans and also the trained model) required for the notebooks can be downloaded from [Zenodo](https://zenodo.org/record/6091361#.Ygu4gIzMI5k. The model can either be trained in the Jupyterlab or can be sent to Galaxy's cluster for remote processing. After remote processing, the created datasets such as the trained model become available in the new Galaxy history.
```
Comment (hexylena) [inline topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:84]:
```suggestion
In this tutorial, we will use the dataset of CTs scans and their respective masks to train an Unet neural network model. The model learns to map the infected regions in the CT scans to their masks. For prediction, the trained model is given unseen CT scans and for each CT scan, it predicts infected regions or masks. For this experiment, we will use Jupyterlab for cloning the notebooks from Github that contain scripts for downloading data, creating and training Unet model and prediction tasks. Data (CT scans and also the trained model) required for the notebooks can be downloaded from [Zenodo](https://zenodo.org/record/6091361). The model can either be trained in the Jupyterlab or can be sent to Galaxy's cluster for remote processing. After remote processing, the created datasets such as the trained model become available in the new Galaxy history.
```

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+
+{:.no_toc}
```
Comment (hexylena) [inline topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md:181]:
```suggestion
```

**Review by hexylena — COMMENTED** [review]
overall formatting and image captions look ok! Thanks @anuprulez

**Comment by kxk302** [comment]
Hi @anuprulez 

I am trying to view your tutorial via gitpod (https://training.galaxyproject.org/training-material/topics/contributing/tutorials/gitpod/tutorial.html). If you look at instructions under 'Already have a fork of the GTN?'
you will see that your fork must have a main branch (not master) and it must be up-to-date. Could you please rename your master branch to main and make sure its up-to-date, so I can use gitpod to view your tutorial? Thanks!

**Comment by hexylena** [comment]
@kxk302 if you check the very bottom left of gitpod you can change branches there to the one this PR is coming from.

image

**Comment by anuprulez** [comment]
@kxk302 I have updated the "master" branch in my fork to "main" (https://github.com/anuprulez/training-material) from which this current feature branch was created and this feature branch is also up-to-date with the `galaxyproject/training-material:main.`
I guess this should work :+1:

**Comment by kxk302** [comment]
> @kxk302 I have updated the "master" branch in my fork to "main" (https://github.com/anuprulez/training-material) from which this current feature branch was created and this feature branch is also up-to-date with the `galaxyproject/training-material:main.` I guess this should work 👍

Thanks @anuprulez !

**Comment by anuprulez** [comment]
@kxk302 was it possible for you to install the branch? thanks but I am not able to see your review comments here

**Comment by kxk302** [comment]
> @kxk302 was it possible for you to install the branch? thanks but I am not able to see your review comments here

@anuprulez if you go tutorial.md in Files Changed tab, you should see my embedded comments. Let me know if you have any issues

**Comment by anuprulez** [comment]
@kxk302 I am not able to see your comments. Here is the link to the file that I am looking at: https://github.com/galaxyproject/training-material/pull/3278/files?file-filters%5B%5D=.md&show-viewed-files=true&w=1

md_file

Also, `Show comments` filter is checked for this file but I don't see any comment. Do I miss something? Can you provide the line number of one of your comments? Thanks

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+
+## Protein 3D structure
+Understanding the structure of proteins provides insights into their functions. But, only a few proteins have known structures out of billions of known proteins. To advance into the direction of predicting their 3D structures only from their sequences, [AlphaFold2](https://www.nature.com/articles/s41586-021-03819-2) has made a breakthrough to predict their structures with high accuracy. However, the database that it uses are large, approximately 2 TBs, and are hard to manage. Therefore, to make it more accessible, a few approaches such as [ColabFold](https://www.biorxiv.org/content/10.1101/2021.08.15.456425v1.full.pdf), has been developed that consumes significantly less memory, much faster and produces 3D structures with similar accuracy. We will look at, in later sections, how we can use ColabFold to predict 3D structure of a protein sequence using the Jupyterlab infrastructure in Galaxy.
```
[the remark this answers is not in the document]
Comment (anuprulez) [reply topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md]: I have moved the references to the bottom of the tutorial. Thanks!

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+
+![Normal and Covid CT scans](../../images/normal_covid_ct_scans.jpg "(Left) Lungs CT scan for a normal person and (right) lungs CT scan for a person having COVID-19")
+
+
+## CT scans and masks
+Figure 4 shows CT scans of infected lungs (top row) and borders around the infected regions have been drawn with red color (middle row). In the last row respective masks, from the CT scans, have been taken out denoting the infected regions. These masks are the "segmented" regions from the corresponding CT scans. While creating the dataset for training the deep learning model (Unet), the CT scans are the data points and its "known" respective masks of infected regions become their [labels](https://www.sciencedirect.com/science/article/pii/S2666990021000069).
```
[the remark this answers is not in the document]
Comment (anuprulez) [reply topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md]: Unet has a small section in the tutorial. Thanks!

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+
+![Image segmentation for a cat](../../images/cat_segmentation.png "Image segmentation for cat. The violet region of the middle plot denotes the cat and the outline of the violet section denoted by the yellow color is the cat's mask")
+
+
+## COVID CT scan
+Segmenting lung CT scans task to locate infected regions has been actively proposed to augment the RT-PCR testing for initial screening of COVID infection in humans. Deep learning has been used to predict these regions with high accuracy in studies such as [Inf-net: Automatic COVID-19 lung infection segmentation from CT images](https://ieeexplore.ieee.org/document/9098956), [JCS: An explainable COVID-19 diagnosis system by joint classification and segmentation](https://arxiv.org/abs/2004.07054) and many more. In Figure 3, the differences between the CT scans of a normal person and a person suffering from COVID-19 can be seen. The regions of lungs marked by white patches are infected by [COVID-19](https://www.sciencedirect.com/science/article/pii/S2666990021000069).
```
[the remark this answers is not in the document]
Comment (anuprulez) [reply topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md]: Moved the references to the bottom.

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+
+![Elyra AI pipeline](../../images/elyra_ai_pipeline.png "An example ML pipeline created using Elyra AI in a notebook. Each connected block of the pipeline is a different notebook")
+
+There are several additional packages suited for performing machine learning tasks such as [Open-CV](https://github.com/opencv/opencv) and [Scikit-Image](https://scikit-image.org/) for image processing, [NiBabel](https://nipy.org/nibabel/) package for processing images files. These packages are available to researchers and users in this notebook. Many other features such as GPU utilization dashboards for monitoring the GPU usage and system memory utilization, voila for rendering output cells of a notebook in a separate tab hiding all code cells, interactive bqplots are integrated into the Jupyterlab notebook for researchers and users to manage their projects more efficiently.
```
[the remark this answers is not in the document]
Comment (anuprulez) [reply topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md]: done!

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+# Reproduce results from recent publications using Jupyterlab infrastructure in Galaxy
+In this tutorial, we will a few features of Jupyterlab to create and train a Unet deep learning model and then, predict segmented regions in COVID CT scans using the Unet trained model. In addition, we will predict 3D structure of protein sequence using ColabFold, a faster implementation of AlphaFold2.
```
[the remark this answers is not in the document]
Comment (anuprulez) [reply topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md]: updated the word, thanks!

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+> 
+> ![Galaxy history](../../images/finished_history_remote_ai.png "Galaxy history showing finished datasets after remote training on a Galaxy cluster")
+> 
+> **Note**: the training may take longer depending on how busy Galaxy's queueing is as it sends the training task to be done on a Galaxy cluster. Therefore, this feature should be used when the training task is expected to run for at least a couple of hours. The training time is higher because a large Docker container is downloaded on the assigned cluster and only then, the training task can proceed.
```
[the remark this answers is not in the document]
Comment (anuprulez) [reply topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md]: Done!

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
... The ready-to-use infrastructure that we used here for developing an AI algorithm and training a model proves convenient for making quick prototypes, accelerated by GPU, in the data science field without caring about the storage and compute resources. The possibility of remote execution of long-running training tasks obviates keeping the notebook or computer open till the training finishes and at the same time, the datasets and models stay safe in Galaxy history. The usage of `h5` files for input datasets to a neural network solves the problem of varying data formats across multiple fields that use AI algorithms extensively for predictive tasks. Using `onnx` for creating, storing and sharing AI models makes it easier to handle such a model as one compact file. Git integration makes it easy to maintain an entire repository directly in the notebook. The entire environment can be shared with other researchers only via sharing the URL. Bioblend connects this infrastructure to Galaxy which makes its histories, workflows and tools accessible directly via the notebook. Overall, this infrastructure would prove useful for researchers working to create quick prototypes of their applications.
```
[the remark this answers is not in the document]
Comment (anuprulez) [reply topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md]: yes, now it has two paragraphs, thanks!

**Comment by anuprulez** [comment]
@kxk302 thanks for reviewing the tutorial. I have added fixes for the review comments. I have added smaller sections for those that were bigger and organised them as more coherent sections. Would be happy to know your further suggestions.

File: topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
...ing labels, validation data, validation labels, test data and test labels. These sub-datasets are stored in different variables after reading the original "h5" file once. For training, we only need these datasets/matrices - training data, training labels, validation data and validation labels. The matrices, test data and test labels, are used for prediction. We use "h5" format for storing and retrieving datasets as all AI algorithms need input datasets in the form of matrices. Since, in the field of AI, there are many different types of datasets such as images, sequences, real numbers and so on, therefore, to converge all these different forms of datasets to one format, we use "h5" to store matrices. In any AI analysis, different forms of datasets can be stored as `h5` files. For image segmentation task, we also saved all the input datasets/matrices to the deep learning model as sub-datasets in one "h5" file so that it can be easily created, stored, downloaded and used. This step may take a few minutes as it downloads around 450 MB of data from Zenodo. Once datasets are downloaded to the notebook, we can move to the next step of creating a deep learning model and start training it.
```
[the remark this answers is not in the document]
Comment (kxk302) [reply topics/statistics/tutorials/gpu_jupyter_lab/tutorial.md]: Thanks Anup, they look very good

**Comment by kxk302** [comment]
@anuprulez the changes look great. Thank you!

**Comment by anuprulez** [comment]
can this be merged once the build passes? Thanks!

**Comment by bgruening** [comment]
I'm halfway through ...

**Comment by anuprulez** [comment]
> I'm halfway through ...

Please let me know if you have any further comments to be fixed/added :)

**Review by bgruening — APPROVED** [review]
This is a great first version @anuprulez 
Thanks a lot!

**Comment by anuprulez** [comment]
Thanks to all @bgruening @hexylena @kxk302

**Pull request merged**